cpg_workflows/large_cohort/combiner.py

```diff
@@ -9,23 +9,123 @@
+from cpg_utils.hail_batch import get_batch, init_batch
+from cpg_workflows.targets import Cohort, SequencingGroup
+from cpg_workflows.utils import can_reuse, slugify, to_path
+from metamist.graphql import gql, query
+
+if TYPE_CHECKING:
+    from graphql import DocumentNode
+
+    from hail.vds.combiner.variant_dataset_combiner import VariantDatasetCombiner
+    from hail.vds.variant_dataset import VariantDataset
+
+
+def _initalise_combiner_job(cohort: Cohort) -> PythonJob:
+    j: PythonJob = get_batch().new_python_job('Combiner', (cohort.get_job_attrs() or {}) | {'tool': 'hail query'})  # type: ignore[reportUnknownArgumentType]
+    j.image(config_retrieve(['workflow', 'driver_image']))
+    j.memory(config_retrieve(['combiner', 'memory']))
+    j.storage(config_retrieve(['combiner', 'storage']))
+
+    # set this job to be non-spot (i.e. non-preemptible)
+    # previous issues with preemptible VMs led to multiple simultaneous QOB groups processing the same data
+    j.spot(config_retrieve(['combiner', 'preemptible_vms'], False))
+    return j
+
+
+def get_vds_ids_output(vds_id: int) -> tuple[str, list[str]]:
+    get_vds_analysis_query: DocumentNode = gql(
+        """
+            query getVDSByAnalysisIds($vds_id: Int!) {
+                analyses(id: {eq: $vds_id}) {
+                    output
+                    sequencingGroups {
+                        id
+                    }
+                }
+            }
+        """,
+    )
+    query_results: dict[str, Any] = query(get_vds_analysis_query, variables={'vds_id': vds_id})
+    vds_path: str = query_results['analyses'][0]['output']
+    vds_sgids: list[str] = [sg['id'] for sg in query_results['analyses'][0]['sequencingGroups']]
+    return (vds_path, vds_sgids)
+
+
+def combiner(cohort: Cohort, output_vds_path: str, save_path: str) -> PythonJob:
+    workflow_config = config_retrieve('workflow')
+    combiner_config = config_retrieve('combiner')
+
+    tmp_prefix_for_withdrawals: str = slugify(
+        f'{cohort.analysis_dataset.tmp_prefix()}{cohort.name}-{workflow_config["sequencing_type"]}-{combiner_config["vds_version"]}',
+    )
+
+    vds_paths: list[str] = []
+    sg_ids_in_vds: list[str] = []
+    sgs_for_withdrawal: list[str] = []
+    new_sg_gvcfs: list[str] = []
+    cohort_sgs: list[SequencingGroup] = cohort.get_sequencing_groups(only_active=True)
+    cohort_sg_ids: list[str] = cohort.get_sequencing_group_ids(only_active=True)
+    gvcf_external_header: str | None = None
+    sequencing_group_names: list[str] | None = None
+
+    if combiner_config.get('vds_analysis_ids', None) is not None:
+        for vds_id in combiner_config['vds_analysis_ids']:
+            tmp_query_res, tmp_sg_ids_in_vds = get_vds_ids_output(vds_id)
+            vds_paths.append(tmp_query_res)
+            sg_ids_in_vds = sg_ids_in_vds + tmp_sg_ids_in_vds
+            sgs_for_withdrawal = [sg for sg in sg_ids_in_vds if sg not in cohort_sg_ids]
```

Comment:
```suggestion
    for vds_id in config_retrieve(['combiner', 'vds_analysis_ids'], []):
        tmp_query_res, tmp_sg_ids_in_vds = get_vds_ids_output(vds_id)
        vds_paths.append(tmp_query_res)
        sg_ids_in_vds = sg_ids_in_vds + tmp_sg_ids_in_vds
    
    sgs_for_withdrawal = [sg for sg in sg_ids_in_vds if sg not in cohort_sg_ids]
```

I'm continuing on my mission to evangelise for `config_retrieve` everywhere. I find the full hierarchy of keys being present in each call easier than remembering which subset of the overall config `combiner_config` represents.

In this situation `sgs_for_withdrawal` is overwritten with each loop anyway, so moving it outside the loop makes sense. If you go this way you can also remove the empty `sgs_for_withdrawal` initialisation above

All personal flavour, feel free to ignore this

Reply:
Not overwriting each time is good, also happy to agree with `config_retrieve`.